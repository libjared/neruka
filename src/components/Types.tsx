type SignedDuration = Duration & {
  negative: boolean;
};

type BaseTenDigit = "0" | "1" | "2" | "3" | "4" | "5" | "6" | "7" | "8" | "9";

type FriendlyDuration = {
  negative: boolean;
  hourTens?: BaseTenDigit;
  hourOnes?: BaseTenDigit;
  minuteTens?: BaseTenDigit;
  minuteOnes?: BaseTenDigit;
  secondTens?: BaseTenDigit;
  secondOnes: BaseTenDigit; // the only required digit
};

export type { SignedDuration, BaseTenDigit, FriendlyDuration };
