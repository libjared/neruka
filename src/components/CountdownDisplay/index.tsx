import { add, differenceInMilliseconds, intervalToDuration, toDate } from "date-fns";
import { useEffect, useState } from "react";

function getCurrentTime(): Date {
  return new Date();
}

type CountdownDisplayProps = {
  targetTime: Date,
  onClick: () => void,
};

function CountdownDisplay({ targetTime, onClick }: CountdownDisplayProps) {
  const [ currentTime, setCurrentTime ] = useState<Date>(getCurrentTime());

  useEffect(() => {
    const interval = setInterval(() => {
      setCurrentTime(getCurrentTime());
    }, 200);

    return () => {
      clearInterval(interval);
    };
  });

  const handleClick = () => {
    onClick();
  };

  const duration = intervalToDuration({
    start: currentTime,
    end: targetTime
  });

  const dur = toFriendlyDuration(duration);

  let str = '';
  if (dur.hourTens !== undefined) { str += `${dur.hourTens}`; }
  if (dur.hourOnes !== undefined) { str += `${dur.hourOnes}h`; }
  if (dur.minuteTens !== undefined) { str += `${dur.minuteTens}`; }
  if (dur.minuteOnes !== undefined) { str += `${dur.minuteOnes}m`; }
  if (dur.secondTens !== undefined) { str += `${dur.secondTens}`; }
  str += `${dur.secondOnes}s`;

  const text = str;

  return (
    <span role="timer" onClick={handleClick}>
      {text}
    </span>
  );
}

type BaseTenDigit = '0' | '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9';

type FriendlyDuration = {
  hourTens?: BaseTenDigit,
  hourOnes?: BaseTenDigit,
  minuteTens?: BaseTenDigit,
  minuteOnes?: BaseTenDigit,
  secondTens?: BaseTenDigit,
  secondOnes: BaseTenDigit, // the only required digit
};

// A whole number is any integer that is zero or positive.
function isWhole(num: number): boolean {
  return Number.isInteger(num) && num >= 0;
}

function toFriendlyDuration(duration: Duration): FriendlyDuration {
  if (duration.days !== undefined && duration.days !== 0) throw new Error("Expected days to be 0 or undefined.");
  if (duration.weeks !== undefined && duration.weeks !== 0) throw new Error("Expected weeks to be 0 or undefined.");
  if (duration.months !== undefined && duration.months !== 0) throw new Error("Expected months to be 0 or undefined.");
  if (duration.years !== undefined && duration.years !== 0) throw new Error("Expected years to be 0 or undefined.");

  if (duration.hours !== undefined && !isWhole(duration.hours)) throw new Error("Expected hours to be a whole number.");
  if (duration.minutes !== undefined && !isWhole(duration.minutes)) throw new Error("Expected minutes to be a whole number.");
  if (duration.seconds !== undefined && !isWhole(duration.seconds)) throw new Error("Expected seconds to be a whole number.");

  const digits: BaseTenDigit[] = new Array(6).fill('0');

  const asBaseTenDigit = (digit: string) => {
    if (digit.length === 1 && digit >= '0' && digit <= '9') {
      return digit as BaseTenDigit;
    }
    return '0'; // this should never happen
  };

  digits[0] = asBaseTenDigit(Math.floor((duration.hours || 0) / 10).toString());
  digits[1] = asBaseTenDigit(((duration.hours || 0) % 10).toString());
  digits[2] = asBaseTenDigit(Math.floor((duration.minutes || 0) / 10).toString());
  digits[3] = asBaseTenDigit(((duration.minutes || 0) % 10).toString());
  digits[4] = asBaseTenDigit(Math.floor((duration.seconds || 0) / 10).toString());
  digits[5] = asBaseTenDigit(((duration.seconds || 0) % 10).toString());

  const friendly: FriendlyDuration = {
    hourTens: digits[0],
    hourOnes: digits[1],
    minuteTens: digits[2],
    minuteOnes: digits[3],
    secondTens: digits[4],
    secondOnes: digits[5],
  };

  if (friendly.hourTens === '0') { delete friendly.hourTens; } else { return friendly; }
  if (friendly.hourOnes === '0') { delete friendly.hourOnes; } else { return friendly; }
  if (friendly.minuteTens === '0') { delete friendly.minuteTens; } else { return friendly; }
  if (friendly.minuteOnes === '0') { delete friendly.minuteOnes; } else { return friendly; }
  if (friendly.secondTens === '0') { delete friendly.secondTens; }
  return friendly;
}

type PreciseDuration = Duration & {
  milliseconds?: number
};

function intervalToPreciseDuration(interval: Interval): PreciseDuration {
  const looseDifference = intervalToDuration(interval);
  const remainingMilliseconds = add(interval.start, { ...looseDifference });
  const milliseconds = differenceInMilliseconds(interval.end, remainingMilliseconds);
  return {
    ...looseDifference,
    milliseconds,
  };
}

function intervalToDurationCeiling(interval: Interval): Duration {
  const start = toDate(interval.start);
  const end = toDate(interval.end);
  const startFrac = start.getTime() % 1000;
  const endFrac = end.getTime() % 1000;

  let fixedEnd = end;
  if (startFrac !== endFrac) {
    // the difference is inexact; round up.
    fixedEnd = add(fixedEnd, { seconds: 1 });
  }
  return intervalToDuration({ start, end: fixedEnd });
}

export default CountdownDisplay;
export { toFriendlyDuration, intervalToPreciseDuration, intervalToDurationCeiling };

export type { FriendlyDuration };
