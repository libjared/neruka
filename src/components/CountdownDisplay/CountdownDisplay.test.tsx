import {
  act,
  fireEvent,
  render,
  RenderResult,
  screen,
} from "@testing-library/react";
import CountdownDisplay, {
  intervalToDurationCeiling,
  toFriendlyDuration,
} from ".";
import { FriendlyDuration } from "../Types";

jest.useFakeTimers();

type SetupResult = RenderResult & {
  advanceClockShort: () => void;
  advanceClockHalfSecond: () => void;
  advanceClockOneSecond: () => void;
  advanceClockToTarget: () => void;
  advanceClock505PastTarget: () => void;
  click: () => void;
  handleClick: jest.Mock<void, void[]>;
  expectedTextInitial: string;
  expectedTextOneSecond: string;
  expectedTextTarget: string;
  expectedText505PastTarget: string;
};

function setup(): SetupResult {
  const fakeCurrentTime = new Date(2023, 1, 15, 0, 0);
  jest.setSystemTime(fakeCurrentTime);
  const targetTime = new Date(2023, 1, 15, 11, 0);
  const handleClick = jest.fn<void, void[]>();
  const utils: RenderResult = render(
    <CountdownDisplay targetTime={targetTime} onClick={handleClick} />
  );
  const advanceClockShort = () => {
    act(() => {
      jest.advanceTimersByTime(199);
    });
  };
  const advanceClockHalfSecond = () => {
    act(() => {
      jest.advanceTimersByTime(500);
    });
  };
  const advanceClockOneSecond = () => {
    act(() => {
      jest.advanceTimersByTime(1000);
    });
  };
  const advanceClockToTarget = () => {
    act(() => {
      const hrs = 11;
      const mins = 0;
      const secs = 0;
      jest.advanceTimersByTime(((hrs * 60 + mins) * 60 + secs) * 1000);
    });
  };
  const advanceClock505PastTarget = () => {
    act(() => {
      const hrs = 11;
      const mins = 5;
      const secs = 5;
      jest.advanceTimersByTime(((hrs * 60 + mins) * 60 + secs) * 1000);
    });
  };
  const click = () => {
    fireEvent.click(screen.getByRole("timer"));
  };

  return {
    ...utils,
    advanceClockShort,
    advanceClockHalfSecond,
    advanceClockOneSecond,
    advanceClockToTarget,
    advanceClock505PastTarget,
    click,
    handleClick,
    // TODO is it necessary to include these as utils
    expectedTextInitial: "11h00m00s",
    expectedTextOneSecond: "10h59m59s",
    expectedTextTarget: "0s",
    expectedText505PastTarget: "-5m05s",
  };
}

it("renders the correct text", () => {
  const { expectedTextInitial } = setup();
  expect(screen.getByText(expectedTextInitial)).toBeInTheDocument();
});

// it("renders the correct text B", () => {
//   const targetTime = new Date(2023, 1, 15, 11, 0);
//   const handleClick = jest.fn<void, void[]>();
//   const utils: RenderResult = render(
//     <CountdownDisplay targetTime={targetTime} onClick={handleClick} />
//   );
//   expect();
// });

it("calls handleClick when clicked", () => {
  const { click, handleClick } = setup();
  expect(handleClick).toBeCalledTimes(0);
  click();
  expect(handleClick).toBeCalledTimes(1);
});

it("matches snapshot", () => {
  const { asFragment } = setup();
  expect(asFragment()).toMatchSnapshot();
});

describe("when a short time passes", () => {
  it("matches snapshot", () => {
    const { asFragment, advanceClockShort } = setup();
    advanceClockShort();
    expect(asFragment()).toMatchSnapshot();
  });

  it("does not change the text", () => {
    const { advanceClockShort, expectedTextInitial } = setup();
    advanceClockShort();
    expect(screen.getByText(expectedTextInitial)).toBeInTheDocument();
  });
});

describe("when a half-second passes", () => {
  it("does not change the text because it rounds up", () => {
    const { advanceClockHalfSecond, expectedTextInitial } = setup();
    advanceClockHalfSecond();
    expect(screen.getByText(expectedTextInitial)).toBeInTheDocument();
  });
});

describe("when a second passes", () => {
  it("matches snapshot", () => {
    const { asFragment, advanceClockOneSecond } = setup();
    advanceClockOneSecond();
    expect(asFragment()).toMatchSnapshot();
  });

  it("changes the text", () => {
    const { advanceClockOneSecond, expectedTextOneSecond } = setup();
    advanceClockOneSecond();
    expect(screen.getByText(expectedTextOneSecond)).toBeInTheDocument();
  });
});

describe("when the target time is exactly now", () => {
  it("shows 0s", () => {
    const { advanceClockToTarget, expectedTextTarget } = setup();
    advanceClockToTarget();
    expect(screen.getByText(expectedTextTarget)).toBeInTheDocument();
  });
});

describe("when the target time has passed", () => {
  it("shows a negative", () => {
    const { advanceClock505PastTarget, expectedText505PastTarget } = setup();
    advanceClock505PastTarget();
    expect(screen.getByText(expectedText505PastTarget)).toBeInTheDocument();
  });
});

describe("toFriendlyDuration", () => {
  it("returns 0s when nothing is passed", () => {
    expect(toFriendlyDuration({ negative: false })).toEqual<FriendlyDuration>({
      negative: false,
      secondOnes: "0",
    });
  });

  it("rejects durations with fractional members", () => {
    expect(() => {
      toFriendlyDuration({ negative: false, hours: 0.5, minutes: 13 });
    }).toThrowError("Expected hours to be a whole number.");
  });

  it("rejects durations with negative members", () => {
    expect(() => {
      toFriendlyDuration({ negative: false, hours: 1, minutes: -13 });
    }).toThrowError("Expected minutes to be a whole number.");
  });

  it("rejects durations with fractional negative members", () => {
    expect(() => {
      toFriendlyDuration({ negative: false, seconds: 1 / -2 });
    }).toThrowError("Expected seconds to be a whole number.");
  });

  it("rejects -0s", () => {
    expect(() => {
      toFriendlyDuration({ negative: true });
    }).toThrowError("Expected duration not to be negative 0.");
  });

  it("returns 15s", () => {
    expect(
      toFriendlyDuration({ negative: false, hours: 0, minutes: 0, seconds: 15 })
    ).toEqual<FriendlyDuration>({
      negative: false,
      secondTens: "1",
      secondOnes: "5",
    });
  });

  it("returns 80h05m40s", () => {
    expect(
      toFriendlyDuration({
        negative: false,
        hours: 80,
        minutes: 5,
        seconds: 40,
      })
    ).toEqual<FriendlyDuration>({
      negative: false,
      hourTens: "8",
      hourOnes: "0",
      minuteTens: "0",
      minuteOnes: "5",
      secondTens: "4",
      secondOnes: "0",
    });
  });

  it("returns -15s", () => {
    expect(
      toFriendlyDuration({ negative: true, hours: 0, minutes: 0, seconds: 15 })
    ).toEqual<FriendlyDuration>({
      negative: true,
      secondTens: "1",
      secondOnes: "5",
    });
  });

  it("returns 0s", () => {
    expect(
      toFriendlyDuration({ negative: false, hours: 0, minutes: 0, seconds: 0 })
    ).toEqual<FriendlyDuration>({
      negative: false,
      secondOnes: "0",
    });
  });
});

describe("intervalToDurationCeiling", () => {
  const base = 1677104120048;

  it("rounds up", () => {
    const result = intervalToDurationCeiling({
      start: new Date(base),
      end: new Date(base + 17600105),
    });
    expect(result).toEqual({
      negative: false,
      hours: 4,
      minutes: 53,
      seconds: 21,
    });
  });

  it("doesnt round when the difference is exact", () => {
    const result = intervalToDurationCeiling({
      start: new Date(base),
      end: new Date(base + 17600000),
    });
    expect(result).toEqual({
      negative: false,
      hours: 4,
      minutes: 53,
      seconds: 20,
    });
  });

  it("returns a negative duration when past", () => {
    const result = intervalToDurationCeiling({
      start: new Date(base),
      end: new Date(base - (104000000 - 86400000)),
    });
    expect(result).toEqual({
      negative: true,
      hours: 4,
      minutes: 53,
      seconds: 20,
    });
  });

  function testRounding(msAfter: number, expectedSeconds: number) {
    it(`returns ${expectedSeconds}s when ${(msAfter / 1000.0).toFixed(
      3
    )}s after`, () => {
      const result = intervalToDurationCeiling({
        start: new Date(base),
        end: new Date(base - msAfter),
      });
      expect(result).toEqual({
        negative: expectedSeconds < 0,
        hours: 0,
        minutes: 0,
        seconds: Math.abs(expectedSeconds),
      });
    });
  }

  describe("returns correctly-rounded durations at precise moments", () => {
    testRounding(-1001, 2);
    testRounding(-1000, 1);
    testRounding(-999, 1);
    testRounding(-1, 1);
    testRounding(0, 0);
    testRounding(1, 0);
    testRounding(999, 0);
    testRounding(1000, -1);
    testRounding(1001, -1);
  });
});
