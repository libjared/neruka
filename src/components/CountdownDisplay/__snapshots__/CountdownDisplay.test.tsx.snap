// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <audio
    src="fifteen minutes to go.ogg"
    title="announcer"
  />
  <span
    role="timer"
  >
    11h00m00s
  </span>
</DocumentFragment>
`;

exports[`when a second passes matches snapshot 1`] = `
<DocumentFragment>
  <audio
    src="fifteen minutes to go.ogg"
    title="announcer"
  />
  <span
    role="timer"
  >
    10h59m59s
  </span>
</DocumentFragment>
`;

exports[`when a short time passes matches snapshot 1`] = `
<DocumentFragment>
  <audio
    src="fifteen minutes to go.ogg"
    title="announcer"
  />
  <span
    role="timer"
  >
    11h00m00s
  </span>
</DocumentFragment>
`;
